### PR TITLE
Composable annotation visualization 

### PR DESCRIPTION
**Pull request by eddyxu**
With the success of `Displayable` image and such types, this PR proposes a composable visualization interface, to make visualize annotations in video, image, sensor data easier in notebook and other environment.

In summary, this PR introduces three orthogonal components:

- `Drawable`, an interface to indicate that this components can be visualized. More concretely, it can be an image, an label, bounding box, segmentations and etc.
- `Draw`,  a container that contains an array of `Drawable`s. As a displayable in Jupyter notebook, Draw will render all drawbles in one single frame (i.e., one image or one lidar scan).
- `Render`, the logic to render the Drawable stack on each other, i.e., rendering bounding box and segmentation over each image. More concretely, a Render can be PIL render, SVG render or even render the `Draw` into json. 

The designed interface:s

```python
class Draw(Displayable, ToDict):
    def draw(self, drawable: Drawable) -> Draw:
         """Can chain multiple drawable together"""
         pass
   
    def __add__(self, other) -> Draw:
         return self.draw(other)

    def display(self): # Can be visualized in notebook directly
       pass

    def to_dict(self):  # Can be used to integrate with other visualization libraries
        pass

class Drawable(...):
    def render(self, render: Render, **kwargs):
         pass

class Render(...):
     def text(...): pass
     def rectangle(...): pass
     def polygon(...): pass

class PILRender(Render): pass
class SVGRender(Render): pass
class JSONRender(Render): pass
```

Similar to Apache Beam / numpy , syntax sugar will be provided to make the rendering annotations much easier, for example, you can use (`+`) to chain images with all 2d annotations (i.e., `box2d`, `mask`, `label`):

```python

img = Image(uri="s3://....")
box = Box2d(xmin=10, ymin=20, xmax=30, ymax=40)
# Render image with 3 bounding boxes
img + box1 + box2 + box3
# Or
img.draw(box1).draw(box2).draw(box3)
```

Moreover, a special `Drawable` -- `Style` is designed to make visualization customizable and composable:

```python
class Style(Drawable):
    def __call__(self, inner: Drawable): pass
    def render(self, render, **kwags): pass

cat_style = rikai.viz.Style(color="yellow",  width=4)
# Render with style
img + cat_style(cat_box) + human_style(human_box) ...
```

We will also override `@` to make style easy to use:

```python
img + cat_box@{color="yellow", width=4}
```

We can put arbitrary logic in `Style` as well:

```python

class LabelStyleMap:
   def __init__(self, color_map):
       self.color_map = color_map

   def render(self, render, **kwargs):
       self.inner.render(render, color=self.color_map[self.inner.label_id])
```
Screen Shot 2021-12-21 at 8 25 31 PM

### CODE REVIEW

File: python/rikai/mixin.py [hunk 1]
```diff
@@ -63,6 +63,26 @@ def display(self, **kwargs) -> "IPython.display.DisplayObject":
...
 
+class Drawable(ABC):
+    """Mixin for a class that is drawable"""
+
+    @abstractmethod
+    def render(self, render: "rikai.viz.Render", **kwargs) -> None:
+        """Render the object using render."""
+
+    def __matmul__(self, style: Union[dict, "rikai.viz.Style"]) -> Drawable:
+        from rikai.viz import Style
```
Comment (changhiskhan) [inline python/rikai/mixin.py:74]: this is to prevent circular import?

File: python/rikai/types/vision.py [hunk 2]
```diff
@@ -209,6 +213,15 @@ def _repr_png_(self):
     def __eq__(self, other) -> bool:
         return isinstance(other, Image) and super().__eq__(other)
 
+    def __or__(self, other: Drawable) -> Draw:
+        """Override ``&`` operator to chain images with
```
Comment (changhiskhan) [inline python/rikai/types/vision.py:217]: |  operator right?

File: python/rikai/types/vision.py [hunk 3]
```diff
@@ -209,6 +213,15 @@ def _repr_png_(self):
...
+        visualization components.
+        """
+        return self.draw(other)
+
+    def render(self, render: "rikai.viz.Render", **kwargs) -> None:
```
Comment (changhiskhan) [inline python/rikai/types/vision.py:222]: Should Image really be a Drawable?

File: python/rikai/viz.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    """
+
+    def __init__(self, **kwarg):
+        self.kwargs = kwarg
+        self.inner = None  # type: Optional[Drawable]
+
+    def __repr__(self):
+        return f"style({self.kwargs})"
+
+    def __call__(self, inner: Drawable) -> Drawable:
```
Comment (changhiskhan) [inline python/rikai/viz.py:49]: if we have say a colormap style, can it be applied to a list of Drawables? Or should we say you have to do [style(x) for x in drawables]?

File: python/rikai/viz.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        # Make a copy of Style so the same style can be applied
+        # to multiple drawables
+        s = Style(**self.kwargs)
+        s.inner = inner
+        return s
+
+    def render(self, render: Render, **kwargs):
```
Comment (changhiskhan) [inline python/rikai/viz.py:56]: when would the user need to supply kwargs in the constructor and also in the render call?

File: python/rikai/viz.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+class Draw(Displayable, ABC):
+    """Draw is a container that contain the elements for visualized lazily."""
+
+    def __init__(self):
+        self.layers = []
+
+    def __repr__(self):
+        first_layer = self.layers[0] if self.layers else "N/A"
+        return f"Draw({first_layer})"
```
Comment (changhiskhan) [inline python/rikai/viz.py:71]: why not do all of the layers?

File: python/rikai/viz.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        if not isinstance(layer, Drawable):
+            raise ValueError(f"{layer} must be a Drawable")
+        self.layers.append(layer)
+        return self
+
+    def __or__(self, other: Drawable) -> Draw:
+        return self.draw(other)
+
+
+class Render(ABC):
```
Comment (changhiskhan) [inline python/rikai/viz.py:83]: The naming is somewhat confusing since Draw and Render are close in meaning.
Maybe call this `Renderer` instead of Render?

**Review by changhiskhan — COMMENTED** [review]
Can you give an example of say how we'd draw boxes with their labels styled by a color map by label?

File: python/rikai/mixin.py (hunk 1 above)
Comment (eddyxu) [reply python/rikai/mixin.py:74]: yes

**Comment by eddyxu** [comment]
> Can you give an example of say how we'd draw boxes with their labels styled by a color map by label?

So it requires additional implementation, some pseudo code  along the line:

```python

from rikai.viz import Style

class StyleMap(Style):
    def __init__(self, style_map: dict[str, Style]):
         super().__init__()
         self.style_map = style_map

    def __call__(self, inner: LabelledDrawable):
         super().__call__(inner)

    def render(self, render, **kwargs):
          style = self.style_map[self.inner.label_id](self.inner)
          return style.render(render, **kwargs)


style_map = StyleMap({ label: Style(color=color) for label, color in zip(labels, color_plates) })
img | style_map(box1) | style_map(box2)
```

File: python/rikai/viz.py (hunk 5 above)
Comment (eddyxu) [reply python/rikai/viz.py:56]: i should make this a private/protected methods tho. This method is not user-facing interface.

File: python/rikai/viz.py (hunk 7 above)
Comment (eddyxu) [reply python/rikai/viz.py:83]: sgtm

File: python/rikai/viz.py (hunk 4 above)
Comment (eddyxu) [reply python/rikai/viz.py:49]: Answered below

File: python/rikai/types/vision.py (hunk 3 above)
Comment (eddyxu) [reply python/rikai/types/vision.py]: Good question. lets not make `Image` a `Drawable` for now.

File: python/rikai/types/vision.py (hunk 2 above)
Comment (eddyxu) [reply python/rikai/types/vision.py]: Fixed.

File: python/rikai/viz.py (hunk 5 above)
Comment (eddyxu) [reply python/rikai/viz.py]: Renamed this method to `_render(..)`

File: python/rikai/viz.py (hunk 6 above)
Comment (eddyxu) [reply python/rikai/viz.py:71]: Most of the boxes and masks (segments) do not have much readable meaning via `__repr__`

**Comment by changhiskhan** [comment]
> > Can you give an example of say how we'd draw boxes with their labels styled by a color map by label?
> 
> So it requires additional implementation, some pseudo code along the line:
> 
> ```python
> from rikai.viz import Style
> 
> class StyleMap(Style):
>     def __init__(self, style_map: dict[str, Style]):
>          super().__init__()
>          self.style_map = style_map
> 
>     def __call__(self, inner: LabelledDrawable):
>          super().__call__(inner)
> 
>     def render(self, render, **kwargs):
>           cur_style = self.style_map[self.inner.label_id](self.inner)
>           return cur_style.render(render, **kwargs)
> 
> 
> style_map = StyleMap({ label: Style(color=color) for label, color in zip(labels, color_palatte) })
> img | style_map(box1) | style_map(box2)
> ```

In the Coco example you don't have a definite number of boxes but instead you have a list of boxes. In this scenario, would it work like `img | list_of_boxes` ? If so how would you apply the style_map? `img | style_map(list_of_boxes)` or `img | [style_map(x) for x in list_of_boxes]` ?

**Review by changhiskhan — APPROVED** [review]
lgtm. This would be a lot more useful for the coco dataset if it was easier to handle:
1. Automatically placing the labels
2. A list of boxes rather than just individual ones
3. applying stylemap to the list altogether.

**Comment by eddyxu** [comment]
For 

> A list of boxes rather than just individual ones
> applying stylemap to the list altogether.

Yes, we can just make `Draw.draw()` interface to take a list of `Drawable` that could work.

like

```python
class Draw:
    def draw(self, drawable: Union[Drawble, Iterator[Drawable]]) -> Draw:
          pass

Similar to Style

class Style:
    def __call__(self, inner: Union[Drawable, Iterator[Drawable]]) -> Drawable:
          pass
```

Automating placing labels is more challenging tho. :) the main concern at the moment is , if we pack label with annotations (i.e., bbox / segmentation), what is the impact of accessing it in Spark SQL, as well as the structural information of the semantic. I think we could come up something in mid term tho.

**Pull request merged**